[FLINK-27146] [Filesystem] Migrate to Junit5

kottmann opened the pull request:
## What is the purpose of the change

This migrates the Filesystem test code to use Junit5 and Assertj. 

## Brief change log

- Migrated the test code to use Junit5 according to the migration documentation
- Also migrated to Assertj (this is in a separate commit and could be dropped or resubmitted in a separate PR)

## Verifying this change

This change will not have any impact on non-test code since only test code was changed. The  change can potentially break tests, but the majority of changes are simple and don't touch the logic.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

ferenc-csaky commented on flink-filesystems/flink-hadoop-fs/src/test/java/org/apache/flink/runtime/fs/hdfs/HadoopFsFactoryTest.java:
```diff
@@ -46,14 +45,14 @@ public void testCreateHadoopFsWithoutConfig() throws Exception {
         try {
             factory.create(uri);
-            fail("should have failed with an exception");
+            Assertions.fail("should have failed with an exception");
```
Use AssertJ's `assertThatThrownBy` instead of try/catch.

ferenc-csaky commented on flink-filesystems/flink-hadoop-fs/src/test/java/org/apache/flink/runtime/fs/hdfs/HadoopFreeTests.java:
Use AssertJ's `assertThatThrownBy` instead of try/catch and `fail()`.

ferenc-csaky commented on flink-filesystems/flink-s3-fs-presto/src/test/java/org/apache/flink/fs/s3presto/PrestoS3FileSystemITCase.java:
```diff
@@ -25,19 +25,15 @@
-import static org.junit.Assert.assertFalse;
-import static org.junit.Assert.fail;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.fail;
```
Use AssertJ's `assertThatThrownBy` instead of try/catch `fail()`.

ferenc-csaky reviewed (changes requested):
Thank you for your contribution! If you have some time, pls. review my comments.

In general, pls. use AssertJ for any kind of asserts. The community agreed on this and it is part of the [JUnit5 migration plan](https://docs.google.com/document/d/1514Wa_aNB9bJUen4xm5uiuXOooOJTtXqS_Jqk9KJitU/edit#heading=h.n2pt1kpwyals). Currently, there are JUnit5 asserts, some `junit.framework.TestCase.assert*`-like imports (this style is a JUnit3 `TestCase` extension residue), and `org.hamcrest.MatcherAssert.assert*`-like imports. All of those should be replaced by AssertJ assertions.

Some test classes and methods were changed to package-private, some are not. Pls. make those changes on all test classes.

kottmann commented:
@ferenc-csaky  Thank you for your insightful feedback and for directing me towards the assertj decision. I updated my branch by rebasing it onto master and incorporating the changes related to assertj. Additionally, I will make another commit to address the remaining feedback that isn't connected to assertj.

ferenc-csaky commented on flink-filesystems/flink-gs-fs-hadoop/src/test/java/org/apache/flink/fs/gs/utils/ConfigUtilsHadoopTest.java:
```diff
@@ -278,8 +280,8 @@ public void shouldProperlyCreateHadoopConfig() {
-        assertEquals(Collections.EMPTY_MAP, difference.entriesDiffering());
-        assertEquals(Collections.EMPTY_MAP, difference.entriesOnlyOnLeft());
-        assertEquals(Collections.EMPTY_MAP, difference.entriesOnlyOnRight());
+        assertThat(difference.entriesDiffering()).isEqualTo(Collections.EMPTY_MAP);
+        assertThat(difference.entriesOnlyOnLeft()).isEqualTo(Collections.EMPTY_MAP);
+        assertThat(difference.entriesOnlyOnRight()).isEqualTo(Collections.EMPTY_MAP);
```
Use `.isEmpty()` instead of equal check.

ferenc-csaky commented on flink-filesystems/flink-gs-fs-hadoop/src/test/java/org/apache/flink/fs/gs/writer/GSRecoverableFsDataOutputStreamTest.java:
```diff
@@ -133,57 +131,57 @@ public void before() {
         if (empty) {
-            assertEquals(0, position);
-            assertEquals(0, componentObjectCount);
+            assertThat(position).isEqualTo(0);
+            assertThat(componentObjectCount).isEqualTo(0);
```
nit: Use `.isZero()`. Applies to every other zero-check in the file.

ferenc-csaky commented on flink-filesystems/flink-gs-fs-hadoop/src/test/java/org/apache/flink/fs/gs/writer/GSRecoverableFsDataOutputStreamTest.java:
```diff
@@ -133,57 +131,57 @@ public void before() {
+        assertThat(blobStorage.blobs.size()).isEqualTo(0);
         fsDataOutputStream.closeForCommit();
-        assertEquals(1, blobStorage.blobs.size());
+        assertThat(blobStorage.blobs.size()).isEqualTo(1);
```
nit: Use `.isOne()`.

ferenc-csaky commented on flink-filesystems/flink-gs-fs-hadoop/src/test/java/org/apache/flink/fs/gs/writer/GSRecoverableWriterCommitterTest.java:
```diff
@@ -132,28 +131,29 @@ public void after() throws IOException {
 
         // there should be exactly one blob left, the final blob identifier. validate its contents.
-        assertEquals(1, blobStorage.blobs.size());
+        assertThat(blobStorage.blobs.size()).isEqualTo(1);
```
nit: Use `.isOne()`.

ferenc-csaky commented on flink-filesystems/flink-hadoop-fs/src/test/java/org/apache/flink/runtime/fs/hdfs/AbstractHadoopFileSystemITTest.java:
```diff
@@ -95,20 +90,20 @@ public void testSimpleFileWriteAndRead() throws Exception {
 
             // directory empty
-            assertEquals(0, fs.listStatus(directory).length);
+            assertThat(fs.listStatus(directory).length).isEqualTo(0);
```
nit: Use `.isZero()`.

ferenc-csaky commented on flink-filesystems/flink-oss-fs-hadoop/src/test/java/org/apache/flink/fs/osshadoop/writer/OSSRecoverableSerializerTest.java:
```diff
@@ -145,6 +101,17 @@ private static OSSRecoverable createOSSRecoverable(
     }
 
+    private static void assertIsEqualTo(OSSRecoverable actual, OSSRecoverable expected) {
+        Assertions.assertThat(actual.getObjectName()).isEqualTo(expected.getObjectName());
+        Assertions.assertThat(actual.getUploadId()).isEqualTo(expected.getUploadId());
+        Assertions.assertThat(actual.getNumBytesInParts()).isEqualTo(expected.getNumBytesInParts());
+        Assertions.assertThat(actual.getLastPartObject()).isEqualTo(expected.getLastPartObject());
+        Assertions.assertThat(actual.getLastPartObjectLength())
+                .isEqualTo(expected.getLastPartObjectLength());
+        Assertions.assertThat(actual.getPartETags().stream().map(PartETag::getETag).toArray())
+                .isEqualTo(expected.getPartETags().stream().map(PartETag::getETag).toArray());
```
nit: static import for `assertThat`.

ferenc-csaky commented on flink-filesystems/flink-s3-fs-base/src/test/java/org/apache/flink/fs/s3/common/writer/RecoverableMultiPartUploadImplTest.java:
```diff
@@ -123,127 +116,82 @@ public void multiplePartAndObjectUploadsShouldBeReflectedInRecoverable() throws
-            final int partNo, final byte[] content) {
-
-        final TestUploadPartResult expectedCompletePart =
+    private static void assertThatHasMultiPartUploadWithPart(
```
Both custom assert method with a for-loop can be simplified to something like this:
```java
private static void assertThatHasUploadedObject(StubMultiPartUploader actual, byte[] content) {
    TestPutObjectResult expectedIncompletePart =
            createPutObjectResult(TEST_OBJECT_NAME, content);

    assertThat(actual.getIncompletePartsUploaded()).contains(expectedIncompletePart);
}
```

ferenc-csaky commented on flink-filesystems/flink-s3-fs-base/src/test/java/org/apache/flink/fs/s3/common/writer/S3RecoverableFsDataOutputStreamTest.java:
```diff
@@ -240,28 +242,37 @@ public void discardingUnpersistedUploadedData() throws IOException {
+        assertThat(multipartUploadUnderTest.getPublishedContents())
+                .isEqualTo(bytesOf("hello world"));
+
+        Assertions.assertThatThrownBy(
```
nit: static import for `assertThatThrownBy`.

ferenc-csaky reviewed (changes requested):
Thanks for going through the files again! Added some minor improvement suggestions, mostly nits about common practice in the repo.

kottmann commented:
Sorry for the delay on my end, I added a new commit to address all your comments above.

ferenc-csaky reviewed (approved):
LGTM! @Jiabao-Sun if you have some time, can you check this PR?

ferenc-csaky commented on flink-filesystems/flink-gs-fs-hadoop/src/test/java/org/apache/flink/fs/gs/writer/GSRecoverableFsDataOutputStreamTest.java:
```diff
@@ -133,57 +131,57 @@ public void before() {
-            assertEquals(0, position);
-            assertEquals(0, componentObjectCount);
+            assertThat(position).isZero();
+            assertThat(componentObjectCount).isOne();
```
This should be `.isZero()`, and causes the Azure fail. Missed that in my prev comment. Pls. fix this, if tests are green, then LGTM :)

kottmann commented:
@ferenc-csaky Thank you for reviewing my PR and for your helpful feedback! I really appreciate your support in improving it.

ferenc-csaky commented:
@kottmann I was not a committer back then, but if you will have some time and desire to resolve the current conflicts, I still believe this would be a nice improvement and will add it afterwards.

Jiabao-Sun commented:
Thanks @kottmann for this contribution and @ferenc-csaky for the review, 
Sorry for missing the message earlier.
Could you rebase master and resolve the conflicts?

kottmann commented:
Thanks, I rebased it onto master and it should be possible to merge now. Some classes received a bit of JUnit 5 migration from another commit before, and I resolved those conflicts or removed my changes.

ferenc-csaky reviewed (approved):
Thanks for updating, LGTM, will merge in 24h if no objections until then.

ferenc-csaky commented on flink-filesystems/flink-hadoop-fs/src/test/java/org/apache/flink/runtime/fs/hdfs/HadoopRecoverableWriterTest.java:
```diff
@@ -49,12 +48,14 @@ class HadoopRecoverableWriterTest extends AbstractRecoverableWriterTest {
     @BeforeAll
     static void testHadoopVersion() {
-        assumeThat(HadoopUtils.isMinHadoopVersion(2, 6)).isTrue();
+        Assumptions.assumeTrue(HadoopUtils.isMinHadoopVersion(2, 7));
```
The actual min Hadoop version is 2.10 for a while: https://github.com/apache/flink/blob/adb8368985ad20e1872a50939c43488516d18b55/pom.xml#L113

So effectively both will be true, it won't matter.

ferenc-csaky replied:
Although, I think we should keep using AssertJ's `assumeThat`.

Jiabao-Sun commented on flink-filesystems/flink-gs-fs-hadoop/src/test/java/org/apache/flink/fs/gs/GSFileSystemFactoryTest.java:
```diff
@@ -22,10 +22,10 @@
+class GSFileSystemFactoryTest {
 
     @Test
     public void testOverrideStorageRootUrl() {
```
```suggestion
   void testOverrideStorageRootUrl() {
```

Jiabao-Sun commented on flink-filesystems/flink-oss-fs-hadoop/src/test/java/org/apache/flink/fs/osshadoop/HadoopOSSRecoverableWriterExceptionITCase.java:
```diff
@@ -36,14 +36,14 @@
 
-    @BeforeClass
+    @BeforeAll
     public static void checkCredentialsAndSetup() throws IOException {
```
```suggestion
   static void checkCredentialsAndSetup() throws IOException {
```

Jiabao-Sun replied:
```suggestion
   static void checkCredentialsAndSetup() throws IOException {
```

Jiabao-Sun commented on flink-filesystems/flink-oss-fs-hadoop/src/test/java/org/apache/flink/fs/osshadoop/HadoopOSSRecoverableWriterITCase.java:
```diff
@@ -39,13 +39,13 @@
 
-    @BeforeClass
+    @BeforeAll
     public static void checkCredentialsAndSetup() throws IOException {
```
```suggestion
   static void checkCredentialsAndSetup() throws IOException {
```

Jiabao-Sun commented on flink-filesystems/flink-s3-fs-presto/src/test/java/org/apache/flink/fs/s3presto/PrestoS3FileSystemITCase.java:
```diff
@@ -48,20 +44,12 @@
-    public static List<String> parameters() {
-        return Arrays.asList("s3", "s3p");
-    }
+class PrestoS3FileSystemITCase extends AbstractHadoopFileSystemITTest {
```
Will we lose parameter tests? Additionally, can we check if the number of tests is consistent with the master branch?

kottmann replied:
The parameter "schema" wasn't used in the test before, therefore nothing should change by removing it. 

Let me compare the number of tests and then comment here.

Jiabao-Sun reviewed (commented):
Thanks @kottmann for this contribution and @ferenc-csaky for the review.
I left some comments.

kottmann commented:
@Jiabao-Sun Thanks for your review, I addressed your comments in the last commit, for the more general things you mentioned I searched in flink-filesystem for other occurrence of the same pattern and changed it too.

I am not sure about the number of tests running, it appears that the number of tests in CI for my branch is lower than for master, and also the test duration is shorter, so it would be good to understand whats going on before merging this.

kottmann commented:
I tracked a missing test case down to HadoopS3FileSystemBehaviorITCase this test doesn't run on my branch, but does run on the master branch. But this wasn't modified in my PR, do you know why this doesn't run? Are they disabled by default in CI for PRs?

Jiabao-Sun reviewed (commented):
Thanks @kottmann for the quick update.
Overall, it was good for me. I just left a few comments. 
Also, did the CI fail this time because of the changes made?

ferenc-csaky commented on flink-filesystems/flink-azure-fs-hadoop/src/test/java/org/apache/flink/fs/azurefs/AzureBlobRecoverableWriterTest.java:
```diff
@@ -46,8 +45,13 @@ class AzureBlobRecoverableWriterTest extends AbstractRecoverableWriterTest {
-        assumeThat(StringUtils.isNullOrWhitespaceOnly(CONTAINER)).isFalse();
-        assumeThat(StringUtils.isNullOrWhitespaceOnly(ACCESS_KEY)).isFalse();
+        Assumptions.assumeFalse(
+                StringUtils.isNullOrWhitespaceOnly(CONTAINER),
+                "Azure container not configured, skipping test...");
+        Assumptions.assumeFalse(
+                StringUtils.isNullOrWhitespaceOnly(ACCESS_KEY),
+                "Azure access key not configured, skipping test...");
```
I think we should revert this part to AssertJ's `aasumeThat`.

ferenc-csaky reviewed (changes requested):
Thank you @Jiabao-Sun for your thorough review! Added some minor comments myself. 

The force-push compare got me tricked, cause it showed only a couple line changes and I thought my last review is still intact. :)

ferenc-csaky reviewed (approved):
I'm pretty sure that the last failure cause by that timeout in the "connect" test module is completely independent from your changes. Either a flaky test or some other merged commit to master.

Thank you for your responsiveness and quick fixes! Checked through all the changed files, and in overall the current state LGTM. I will merge it tomorrow if no objections until then. cc @Jiabao-Sun

kottmann commented:
@ferenc-csaky Shall I squash the commits, or will you do that when it gets merged?

ferenc-csaky commented:
> @ferenc-csaky Shall I squash the commits, or will you do that when it gets merged?

Squash and merge takes care of that, so it is not necessary I would say.

Jiabao-Sun reviewed (approved):
Thanks @kottmann for the great work and @ferenc-csaky for the review.
LGTM now.

kottmann commented:
@ferenc-csaky  @Jiabao-Sun  Thank you two for your helpful review. If there are any more tickets related to the JUnit 5 migration, I can help with I am happy to take one over.
